geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterElderManager.java

```diff
@@ -0,0 +1,208 @@
+import org.apache.geode.internal.util.concurrent.StoppableReentrantLock;
+
+public class ClusterElderManager {
+  private static final Logger logger = LogService.getLogger();
+
+  private final ClusterDistributionManager clusterDistributionManager;
+  private final StoppableReentrantLock elderLock;
+
+  private ElderState elderState;
+  private volatile boolean elderStateInitialized;
+  private final Supplier<ElderState> elderStateSupplier;
+
+  public ClusterElderManager(ClusterDistributionManager clusterDistributionManager) {
+    this(clusterDistributionManager, () -> new ElderState(clusterDistributionManager));
+  }
+
+  public ClusterElderManager(ClusterDistributionManager clusterDistributionManager,
+      Supplier<ElderState> elderStateSupplier) {
+    this.clusterDistributionManager = clusterDistributionManager;
+    this.elderLock = new StoppableReentrantLock(clusterDistributionManager.getCancelCriterion());
+    this.elderStateSupplier = elderStateSupplier;
+  }
+
+  /**
+   * Based on a recent JGroups view, return a member that might be the next elder.
+   *
+   * @return the elder candidate, possibly this VM.
+   */
+  InternalDistributedMember getElderCandidate() {
+    return getElderCandidates().stream().findFirst().orElse(null);
+  }
+
+  public boolean isPeerless() {
+    List<InternalDistributedMember> candidates = getElderCandidates();
+
+    return candidates.size() == 1 && clusterDistributionManager.getId()
+        .equals(candidates.get(0));
+  }
+
+  List<InternalDistributedMember> getElderCandidates() {
+    List<InternalDistributedMember> theMembers = clusterDistributionManager.getViewMembers();
+
+    return theMembers.stream()
+        .filter(member -> member.getVmKind() != ClusterDistributionManager.ADMIN_ONLY_DM_TYPE)
+        .filter(
+            member -> !clusterDistributionManager.getMembershipManager().isSurpriseMember(member))
+        .collect(Collectors.toList());
+  }
+
+  public InternalDistributedMember getElderId() throws DistributedSystemDisconnectedException {
+    if (clusterDistributionManager.isCloseInProgress()) {
+      throw new DistributedSystemDisconnectedException(
+          LocalizedStrings.DistributionManager_NO_VALID_ELDER_WHEN_SYSTEM_IS_SHUTTING_DOWN
+              .toLocalizedString(),
+          clusterDistributionManager.getRootCause());
+    }
+    clusterDistributionManager.getSystem().getCancelCriterion().checkCancelInProgress(null);
+
+    return getElderCandidate();
+  }
+
+  public boolean isElder() {
+    return clusterDistributionManager.getId().equals(getElderCandidate());
+  }
+
+  public ElderState getElderState(boolean waitToBecomeElder) {
+    if (waitToBecomeElder) {
+      // This should always return true.
+      waitForElder(clusterDistributionManager.getId());
+    }
+
+    if (clusterDistributionManager.getId().equals(getElderCandidate())) {
+      // we are the elder
+      if (this.elderStateInitialized) {
```

Comment:
I think it would be easier to understand if this block, "if (this.elderStateInitialized {...}" was moved into the initializeElderState method.